File: NettyCSharp/com.netty.csharp/client/NettyClientHandler.cs
#region << 版 本 注 释 >>
/*----------------------------------------------------------------
* 类 名 称 ：NettyClientHandler
* 类 描 述 ：处理客户端的channel
* 作    者 ：lucher
* 版 本 号 ：v1.0.0
*******************************************************************
* Copyright @ lucher 2019. All rights reserved.
*******************************************************************
//----------------------------------------------------------------*/
#endregion
using DotNetty.Buffers;
using DotNetty.Transport.Channels;
using EventBus;
using MessagePack;
using System;
using System.Collections.Generic;
using System.Text;

namespace NettyCSharp
{
    public class NettyClientHandler : SimpleChannelInboundHandler<Object>
    {

        protected override void ChannelRead0(IChannelHandlerContext ctx, object msg)
        {
            //String json = msg as String;
            //MessagePackSerializer.;
            //读取到消息，对消息进行解析
            //这样判断感觉有些尴尬，由于对C#不熟，暂时还没有更好的办法，有好的实现方法还望给出宝贵建议
            object[] values = (object[])msg;
            int index = Convert.ToInt32(values[0]);//获取消息类别，第一个字段为EventType，这是约定好的
            String eventType = Enum.GetName(typeof(EventType), index);
            if (eventType.Equals(EventType.TEST_EVENT.ToString())) {//测试类事件
                TestEvent testEvent = MessagePackSerializer.Deserialize<TestEvent>(MessagePackSerializer.Serialize(msg));
                Console.WriteLine("Client接收到消息:" + testEvent);
                SimpleEventBus.GetDefaultEventBus().Post(testEvent.ToString(), TimeSpan.Zero);
            }
            else if (eventType.Equals(EventType.OTHER_EVENT.ToString()))
            {//其他类事件
                Console.WriteLine("Client接收到消息:其他类事件，暂未处理");
            }

        }

        public override void ChannelReadComplete(IChannelHandlerContext context)
        {
            base.ChannelReadComplete(context);
            context.Flush();
            Console.WriteLine("ChannelReadComplete:" + context);
        }

        public override void ChannelRegistered(IChannelHandlerContext context)
        {
            base.ChannelRegistered(context);
            Console.WriteLine("Client ChannelRegistered:" + context);
        }

        public override void ChannelActive(IChannelHandlerContext context)
        {
            Console.WriteLine("Client channelActive:" + context);
            SimpleEventBus.GetDefaultEventBus().Post("建立连接："+context, TimeSpan.Zero);
        }

        public override void ChannelInactive(IChannelHandlerContext context)
        {
            base.ChannelInactive(context);
            Console.WriteLine("Client ChannelInactive:" + context);
            SimpleEventBus.GetDefaultEventBus().Post("连接断开：" + context, TimeSpan.Zero);
        }

        public override void ChannelUnregistered(IChannelHandlerContext context)
        {
            base.ChannelUnregistered(context);
            Console.WriteLine("Client ChannelUnregistered:" + context);
        }

        public override void ExceptionCaught(IChannelHandlerContext context, Exception exception)
        {
            Console.WriteLine("Exception: " + exception);
            context.CloseAsync();
        }
    }
}

File: NettyAndroid/app/src/main/java/com/netty/android/client/NettyClientHandler.java
package com.netty.android.client;

import com.netty.android.coder.MessageConverter;
import com.netty.android.event.EventType;
import com.netty.android.event.TestEvent;

import org.greenrobot.eventbus.EventBus;
import org.msgpack.type.ArrayValue;
import org.msgpack.type.IntegerValue;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;

/**
 * 处理客户端的channel
 *
 * @author lucher
 */
public class NettyClientHandler extends SimpleChannelInboundHandler<ArrayValue> {

    @Override
    protected void channelRead0(ChannelHandlerContext arg0, ArrayValue value) throws Exception {
        // 获取EventType,BaseEvent的第一个字段是EventType，这是约定好的
        int typeIndex = ((IntegerValue) value.get(0)).intValue();
        EventType eventType = EventType.class.getEnumConstants()[typeIndex];
        switch (eventType) {
            case TEST_EVENT:
                TestEvent testEvent = MessageConverter.converter(value, TestEvent.class);
                System.out.println("Client接收到消息:" + testEvent);
                EventBus.getDefault().post(testEvent.toString());

                break;
            case OTHER_EVENT:
                System.out.println("Client接收到消息:其他事件，暂未处理");

                break;
            default:
                break;
        }
    }

    @Override
    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
        super.channelReadComplete(ctx);
        ctx.flush();
    }

    @Override
    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        super.channelRegistered(ctx);
        System.out.println("Client channelRegistered:" + ctx);
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        System.out.println("Client channelActive:" + ctx);
        EventBus.getDefault().post("建立连接:" + ctx);
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        super.channelInactive(ctx);
        System.out.println("Client channelInactive:" + ctx);
        EventBus.getDefault().post("连接断开:" + ctx);
    }

    @Override
    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {// 断开链接事件
        super.channelUnregistered(ctx);
        System.out.println("Client channelUnregistered:" + ctx);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        System.out.println("Client exception：" + cause.getMessage());
        EventBus.getDefault().post("Client exception：" + cause.getMessage());
    }

}